rfcs/2021-10-29-9465-splunk-hec-indexer-acknowledgements.md

```diff
@@ -0,0 +1,399 @@
+{
+    "acks":
+        {
+            "0": true,
+            "1": false,
+            "2": true
+        }
+}
+```
+
+In addition to the overall protocol, there are a few details worth highlighting:
+
+* Authentication tokens
+  * To submit requests to HEC in general, an authentication token must be
+    included in request headers. Indexer acknowledgement is enabled per
+    authentication token in Splunk. The `splunk_hec` source currently supports a
+    list of valid authentication tokens. The `splunk_hec` sinks currently
+    support configuring a single authentication token.
+
+* Splunk channels
+  * Upon enabling indexer acknowledgement, Splunk requires HEC requests to
+    include a channel ID (any valid `guid` value). Querying
+    `/services/collector/ack` also requires a channel ID.
+  * `ackId`’s are assigned per-channel.
+
+* Pending acks
+  * Clients are not required or guaranteed to query for `ackId` status. To avoid
+    running out of memory from `ackId` build up, Splunk offers a few
+    configuration options to limit the number of pending `ackId`’s both overall
+    and per-channel.
+  * Splunk does not explicitly indicate whether event data is dropped or still
+    pending processing. Either of these states will result in `{ack_id}: false`
+    status. Splunk only advises that after a certain amount of time (e.g. 5
+    minutes), the data can be considered dropped.
+
+## Proposal: `splunk_hec` Source Indexer Acknowledgement
+
+### User Experience
+
+Users can configure the `splunk_hec` source with additional indexer
+acknowledgement related settings.
+
+```toml
+[sources.splunk_hec]
+type = "splunk_hec"
+# ...
+acknowledgements.enabled = true
+acknowledgements.max_pending_acks = 10_000_000
+acknowledgements.max_number_of_ack_channel = 1_000_000
+acknowledgements.max_pending_acks_per_channel = 1_000_000
+acknowledgements.ack_idle_cleanup = true
+acknowledgements.max_idle_time = 300
+```
+
+* `acknowledgements.enabled` This controls indexer acknowledgement enablement.
+  Defaults to `false` matching Splunk HEC's opt-in behavior.
+* `acknowledgements.max_pending_acks` With acknowledgements enabled, this
+  controls the maximum number of pending query ackId's overall (across all channels)
+  Defaults to `10_000_000` (Splunk default).
+* `acknowledgements.max_number_of_ack_channel` This controls the max number of
+  channels a client can use with the `splunk_hec` source. Defaults to
+  `1_000_000` (Splunk default).
+* `acknowledgements.max_pending_acks_per_channel` This controls the max number
+  of pending query ackId's per channel. Defaults to `1_000_000` (Splunk
+  default).
+* `acknowledgements.ack_idle_cleanup` This controls whether the `splunk_hec`
+  source will drop channel information (ackId's, statuses) after `max_idle_time`
+  seconds. Defaults to `false` (Splunk default).
+* `acknowledgements.max_idle_time` This controls the max channel idle time
+  before removal. Defaults to `600` seconds (Splunk default).
```

Comment:
Do you have a link to these settings in Splunk's docs? I'm having trouble finding them.

Reply:
Yep, they're at the very bottom of this [Channel limits and indexing status expiration](https://docs.splunk.com/Documentation/Splunk/7.1.2/Data/AboutHECIDXAck#Channel_limits_and_indexing_status_expiration) section. (pasted below)

Parameter | Value type | Default value | Description
-- | -- | -- | --
ackIdleCleanup | boolean | false | When set to true, this parameter causes the server to remove channels that are idle for the number of seconds set in the maxIdleTime setting.
maxIdleTime | integer | 600 | Specifies the maximum number of seconds that channels can be idle before they are removed.

Reply:
Aha, great, thanks!